File: app/models/extension.rb
class Extension < Finance::Base
	
	validates :days, presence: true, numericality: { only_integer: true, greater_than: 0, less_than: 31 } 
	
	belongs_to :loan
	
	before_create do
		if !loan.blank?
			# set own properties
			self.apr = loan.apr * 1.5
			self.interest = loan.principal * self.apr / 365 / 100 * self.days
			self.total = loan.total + self.interest
		end
  end
	
	after_create do
		# update loan
		loan.maturity_date = loan.maturity_date + self.days.days
		loan.total = self.total
		loan.apr = self.apr
		loan.save
	end
	
end

File: app/assets/javascripts/shared/models/Extension.js
(function(){

	var class_name = 'Extension';
	
	Finance[ class_name ] = function()
	{
		Finance.Base.apply( this, arguments );
	}
	
	Finance.register( class_name );

	/*
	 * class properties
	 */
	Finance[ class_name ].prototype.properties =
	[
		'uuid',
		'loan_id',
		'days',
		'apr',
		'interest',
		'total',
	];
	
	/*
	 * loan()
	 */
	Finance[ class_name ].prototype.loan = function()
	{
		return Finance.Loan.find( this.loan_id );
	}
	
	/*
	 * get_total()
	 */
	Finance[ class_name ].prototype.get_total = function()
	{
		var self = this;
		return self.loan().total + self.loan().get_extension_cost( self.days );
	}
	
})();
